boa_engine/src/syntax/parser/expression/primary/class_expression/mod.rs

```diff
@@ -0,0 +1,81 @@
+use crate::syntax::{
+    ast::{Keyword, Node},
+    lexer::TokenKind,
+    parser::{
+        statement::{BindingIdentifier, ClassTail},
+        AllowAwait, AllowYield, Cursor, ParseError, TokenParser,
+    },
+};
+use boa_interner::{Interner, Sym};
+use boa_profiler::Profiler;
+use std::io::Read;
+
+/// Class expression parsing.
+///
+/// More information:
+///  - [ECMAScript specification][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#prod-ClassExpression
+#[derive(Debug, Clone, Copy)]
+pub(super) struct ClassExpression {
+    name: Option<Sym>,
+    allow_yield: AllowYield,
+    allow_await: AllowAwait,
+}
+
+impl ClassExpression {
+    /// Creates a new `ClassExpression` parser.
+    pub(in crate::syntax::parser) fn new<N, Y, A>(name: N, allow_yield: Y, allow_await: A) -> Self
+    where
+        N: Into<Option<Sym>>,
+        Y: Into<AllowYield>,
+        A: Into<AllowAwait>,
+    {
+        Self {
+            name: name.into(),
+            allow_yield: allow_yield.into(),
+            allow_await: allow_await.into(),
+        }
+    }
+}
+
+impl<R> TokenParser<R> for ClassExpression
+where
+    R: Read,
+{
+    type Output = Node;
+
+    fn parse(
+        self,
+        cursor: &mut Cursor<R>,
+        interner: &mut Interner,
+    ) -> Result<Self::Output, ParseError> {
+        let _timer = Profiler::global().start_event("ClassExpression", "Parsing");
+        let strict = cursor.strict_mode();
+        cursor.set_strict_mode(true);
+
+        let token = cursor.peek(0, interner)?.ok_or(ParseError::AbruptEnd)?;
```

Comment:
Looking at the spec, the first token of this parser should be the `class` keyword, right? I guess this should "expect" that keyword.

I guess it's not here because the parser that calls it consumes it before calling it, but I think that should be changed.

Reply:
Yeah that is exactly it. I will open a issue for this, because it will mean replacing `next` with `peek` in multiple places and as a result multiple parsers need to be changed. We have this behaviour in lot's of parsers, so doing this in a dedicated issue seems like the best idea.